text/0000-builder-spec.md

```diff
@@ -0,0 +1,270 @@
+* Buildpacks &rarr; A list of buildpack references, and an ordering of buildpacks
+* Lifecycle &rarr; A version of a lifecycle implementation
+* Stack’s build image &rarr; An environment used at build-time
+* Metadata, describing the builder (description, who it was created by, etc)
+
+Users of `pack` interact with builders in the following ways:
+* `pack build` &rarr; requires a builder to construct an image
+* `set-default-builder <builder>` &rarr; Set default builder used by other commands
+* `inspect-builder <builder>` &rarr; Show information about a builder
+* `suggest-builders` &rarr; Display list of recommended builders
+* `trust-builder <builder>` &rarr; Trust builder
+* `untrust-builder <builder>` &rarr; Stop trusting builder
+* `list-trusted-builders` &rarr; List Trusted Builders
+* `create-builder` &rarr; Create builder image
+
+Users of `pack-orb`, `tekton`, and `kpack` interact with Builders through supplying a builder to use for building applications.
+
+A sample inspected builder (using `pack inspect-builder`), looks like:
+```bash
+ $  pack inspect-builder cnbs/sample-builder:bionic
+Inspecting builder: cnbs/sample-builder:bionic
+
+REMOTE:
+
+Created By:
+  Name: Pack CLI
+  Version: 0.13.1+git-4134cc6.build-1135
+
+Trusted: No
+
+Stack:
+  ID: io.buildpacks.samples.stacks.bionic
+
+Lifecycle:
+  Version: 0.9.1
+  Buildpack APIs:
+    Deprecated: (none)
+    Supported: 0.2, 0.3, 0.4
+  Platform APIs:
+    Deprecated: (none)
+    Supported: 0.3, 0.4
+
+Run Images:
+  cnbs/sample-stack-run:bionic
+
+Buildpacks:
+  ID                            VERSION        HOMEPAGE
+  samples/java-maven            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/java-maven
+  samples/kotlin-gradle         0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/kotlin-gradle
+  samples/ruby-bundler          0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/ruby-bundler
+  samples/hello-universe        0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-universe
+  samples/hello-world           0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-world
+  samples/hello-moon            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon
+
+Detection Order:
+ ├ Group #1:
+ │  └ samples/java-maven@0.0.1
+ ├ Group #2:
+ │  └ samples/kotlin-gradle@0.0.1
+ ├ Group #3:
+ │  └ samples/ruby-bundler@0.0.1
+ └ Group #4:
+    └ samples/hello-universe@0.0.1
+       └ Group #1:
+          ├ samples/hello-world@0.0.1
+          └ samples/hello-moon@0.0.1
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+## Structure of Builders
+### Files/Directories
+A builder is composed of at least the following directories/files:
+```
+- /cnb/buildpacks/...<buildpack ID>/<buildpack version>/
+- /cnb/lifecycle/<lifecycle binaries>
+- /cnb/order.toml
+- /cnb/stack.toml
+- /layers
+- /platform/env/<env-files>
```

Comment:
I don't think we should expect and certainly not require environment variable files to be baked into the builder image. These are supplied by the user for a specific build and are typically mounted in (although `pack` does bake these into an ephemeral builder image).